perception/occupancy_grid_map_outlier_filter/src/occupancy_grid_map_outlier_filter_nodelet.cpp

```diff
@@ -150,43 +150,70 @@ void RadiusSearch2dFilter::filter(
-    xy_cloud->points[i].y = high_conf_xyz_cloud.points[i - low_conf_xyz_cloud.points.size()].y;
+
+  // Low conf cloud check
+  std::vector<int> k_indices_low(low_conf_xy_cloud->points.size());
+  std::vector<float> k_distances_low(low_conf_xy_cloud->points.size());
+  kd_tree_->setInputCloud(low_conf_xy_cloud);
+  for (size_t i = 0; i < low_conf_input.points.size(); ++i) {
+    const float distance = std::hypot(
+      low_conf_xy_cloud->points[i].x - pose.position.x,
+      low_conf_xy_cloud->points[i].y - pose.position.y);
+    const int min_points_threshold = std::min(
+      std::max(static_cast<int>(min_points_and_distance_ratio_ / distance + 0.5f), min_points_),
+      max_points_);
+    const int points_num =
+      kd_tree_->nearestKSearch(i, min_points_threshold, k_indices_low, k_distances_low);
+    if (min_points_threshold <= points_num) {
+      output.points.push_back(low_conf_input.points.at(i));
+    } else {
+      outlier.points.push_back(low_conf_input.points.at(i));
+    }
   }
 
-  std::vector<int> k_indices(xy_cloud->points.size());
-  std::vector<float> k_distances(xy_cloud->points.size());
-  kd_tree_->setInputCloud(xy_cloud);
-  for (size_t i = 0; i < low_conf_xyz_cloud.points.size(); ++i) {
-    const float distance =
-      std::hypot(xy_cloud->points[i].x - pose.position.x, xy_cloud->points[i].y - pose.position.y);
+  // High conf cloud check
+  if (outlier.points.size() == 0) {
+    return;
+  }
+
+  pcl::PointCloud<pcl::PointXY>::Ptr high_conf_xy_cloud(new pcl::PointCloud<pcl::PointXY>);
+  high_conf_xy_cloud->points.resize(high_conf_input.points.size());
+  for (size_t i = 0; i < high_conf_input.points.size(); ++i) {
+    high_conf_xy_cloud->points[i].x = high_conf_input.points[i].x;
+    high_conf_xy_cloud->points[i].y = high_conf_input.points[i].y;
+  }
+
+  std::vector<int> k_indices_high(high_conf_xy_cloud->points.size());
+  std::vector<float> k_distances_high(high_conf_xy_cloud->points.size());
+  kd_tree_->setInputCloud(high_conf_xy_cloud);
+  for (size_t i = 0; i < outlier.points.size(); ++i) {
+    const float distance = std::hypot(
+      high_conf_xy_cloud->points[i].x - pose.position.x,
+      high_conf_xy_cloud->points[i].y - pose.position.y);
     const int min_points_threshold = std::min(
       std::max(static_cast<int>(min_points_and_distance_ratio_ / distance + 0.5f), min_points_),
       max_points_);
     const int points_num =
-      kd_tree_->radiusSearch(i, search_radius_, k_indices, k_distances, min_points_threshold);
+      kd_tree_->nearestKSearch(i, min_points_threshold, k_indices_high, k_distances_high);
 
     if (min_points_threshold <= points_num) {
-      output.points.push_back(low_conf_xyz_cloud.points.at(i));
+      output.points.push_back(outlier.points.at(i));
     } else {
-      outlier.points.push_back(low_conf_xyz_cloud.points.at(i));
+      outlier.points.erase(outlier.points.begin() + i);
```

Comment:
Does it wotk well?
I think you cannot modify loop target data structures (in this case, `outlier.points` ).